paymentsheet/src/main/java/com/stripe/android/paymentsheet/ui/AddPaymentMethodInteractor.kt

```diff
@@ -0,0 +1,234 @@
+        val processing: Boolean,
+        val usBankAccountFormArguments: USBankAccountFormArguments,
+        val linkConfigurationCoordinator: LinkConfigurationCoordinator,
+    )
+
+    sealed class ViewAction {
+        data class OnPaymentMethodSelected(val code: PaymentMethodCode) : ViewAction()
+        data class OnLinkSignUpStateUpdated(val state: InlineSignupViewState) : ViewAction()
+        data class OnFormFieldValuesChanged(
+            val formValues: FormFieldValues?,
+            val selectedPaymentMethodCode: String
+        ) : ViewAction()
+
+        data class ReportFieldInteraction(val code: PaymentMethodCode) : ViewAction()
+    }
+}
+
+internal class DefaultAddPaymentMethodInteractor(
+    private val initiallySelectedPaymentMethodType: PaymentMethodCode,
+    private val linkConfigurationCoordinator: LinkConfigurationCoordinator,
+    private val selection: StateFlow<PaymentSelection?>,
+    private val linkSignupMode: StateFlow<LinkSignupMode?>,
+    private val processing: StateFlow<Boolean>,
+    private val supportedPaymentMethods: List<SupportedPaymentMethod>,
+    private val createFormArguments: (PaymentMethodCode) -> FormArguments,
+    private val formElementsForCode: (PaymentMethodCode) -> List<FormElement>,
+    private val clearErrorMessages: () -> Unit,
+    private val onLinkSignUpStateUpdated: (InlineSignupViewState) -> Unit,
+    private val reportFieldInteraction: (PaymentMethodCode) -> Unit,
+    private val onFormFieldValuesChanged: (FormFieldValues?, String) -> Unit,
+    private val reportPaymentMethodTypeSelected: (PaymentMethodCode) -> Unit,
+    private val createUSBankAccountFormArguments: (PaymentMethodCode) -> USBankAccountFormArguments,
+    dispatcher: CoroutineContext = Dispatchers.Default,
+) : AddPaymentMethodInteractor {
+
+    constructor(sheetViewModel: BaseSheetViewModel) : this(
+        initiallySelectedPaymentMethodType = sheetViewModel.initiallySelectedPaymentMethodType,
+        linkConfigurationCoordinator = sheetViewModel.linkConfigurationCoordinator,
+        selection = sheetViewModel.selection,
+        linkSignupMode = sheetViewModel.linkSignupMode,
+        processing = sheetViewModel.processing,
+        supportedPaymentMethods = sheetViewModel.supportedPaymentMethods,
+        createFormArguments = sheetViewModel::createFormArguments,
+        formElementsForCode = sheetViewModel::formElementsForCode,
+        clearErrorMessages = sheetViewModel::clearErrorMessages,
+        onLinkSignUpStateUpdated = sheetViewModel::onLinkSignUpStateUpdated,
+        reportFieldInteraction = sheetViewModel::reportFieldInteraction,
+        onFormFieldValuesChanged = sheetViewModel::onFormFieldValuesChanged,
+        reportPaymentMethodTypeSelected = sheetViewModel::reportPaymentMethodTypeSelected,
+        createUSBankAccountFormArguments = { USBankAccountFormArguments.create(sheetViewModel, it) }
+    )
+
+    private val coroutineScope = CoroutineScope(dispatcher + SupervisorJob())
```

Comment:
I wonder if we should be "extending" the viewModel scope here, then you wouldn't need to close all the screens when sending a result.

Regardless, probably an idea for a follow up.

Reply:
hmm yeah, that seems like a good idea. I can try that out in a follow up